src/cute_shader/cute_shader.cpp

```diff
@@ -0,0 +1,238 @@
+#include "cute_shader.h"
+#include <stdio.h>
+#include <stdlib.h>
+#include <stdint.h>
+#include <unordered_set>
+#include <unordered_map>
+#include <glslang/Public/ShaderLang.h>
+#include <glslang/Public/ResourceLimits.h>
+#include <StandAlone/DirStackFileIncluder.h>
+#include <SPIRV/GlslangToSpv.h>
+
+#define CUTE_SHADER_GLSL_VERSION 450
+
+namespace cute_shader {
+
+struct BuiltinInclude {
+	const char* content;
+	size_t size;
+};
+
+class Includer: public glslang::TShader::Includer {
```

Comment:
Interesting, is this includer thing a part of the glslang API? I hate it, but looks like it's just a way to use glslang, which is accpetable and I'm glad you hid this all back here in the implementation in an isolated place.

Reply:
Yes, it is glslang API to search for include file. It is passed to glslang:TShader::preprocess. It will call the includer to search for files.

I use that to implement lookup in the VFS, builtin shader include and also automatic include guard.

In the case of duplicated includes, the content is resolved into an empty string.